modules/storages/lib/api/v3/file_links/file_link_representer.rb

```
@@ -0,0 +1,136 @@
+#++
+
+module API
+  module V3
+    module FileLinks
+      class FileLinkRepresenter < ::API::Decorators::Single
+        include API::Decorators::LinkedResource
+        include API::Decorators::DateProperty
+        include ::API::Caching::CachedRepresenter
+
+        property :id
+
+        date_time_property :created_at, writeable: false
+
+        date_time_property :updated_at, writeable: false
+
+        property :originData,
+                 exec_context: :decorator,
+                 getter: ->(*) { make_origin_data(represented) },
+                 setter: ->(fragment:, **) do
+                   parse_origin_data(fragment).each do |attribute, value|
+                     represented[attribute] = value
+                   end
+                 end
+
+        link :self do
+          {
+            href: api_v3_paths.file_link(represented.id)
+          }
+        end
+
+        link :delete, cache_if: -> { user_allowed_to_manage?(represented) } do
+          {
+            href: api_v3_paths.file_link(represented.id),
+            method: :delete
+          }
+        end
+
+        link :creator do
+          {
+            href: api_v3_paths.user(represented.creator_id),
+            title: represented.creator.name
+          }
+        end
+
+        link :originOpen do
+          {
+            href: ::Storages::StorageUrlService.new(represented).call('open').result
+          }
+        end
+
+        link :staticOriginOpen do
+          {
+            href: api_v3_paths.file_link_open(represented.id)
+          }
+        end
+
+        associated_resource :storage
+
+        associated_resource :storageUrl,
+                            skip_render: ->(*) { true },
+                            getter: ->(*) {},
+                            setter: ->(fragment:, **) {
+                              canonical_url = fragment['href'].gsub(/\/+$/, '')
+                              represented.storage = ::Storages::Storage.find_by(host: canonical_url)
```

Depending on the level of error handling you want to provide, it might make sense to have this fall back to something that represents a non existent storage as e.g. done [for work packages](https://github.com/opf/openproject/blob/dev/app/models/work_package/inexistent_work_package.rb). But that will then have to be handled in the contract as well.

Without the mechanism, every non existent storage url provided will lead to `represented.storage.nil?` so the error will be dropped silently.

There is effort involved in setting the aforementioned pattern up though, so it depends on whether it is worth it.

---

NullObject pattern.

We'll then have to test for `storage.is_a?(InexistentStorage)` instead of `storage.nil?` in the contract, which looks similar to me.I'm not sure to get the benefit here. Can you develop, or can we have a call about it?

---

⇒ distinguish between the user supplying a `nil` value, and the user supplying a value that does not exist. If they both resolve to `nil`, then we are unable to distinguish between both. It's really just error handling.

⇒ means the error message needs to be tested and improved if needed.

---

benefit of having a nullobject inheriting from the activerecord: it can store the faulty value(s).

---

Done in e3b7d4b64c.